### PR TITLE
Postgres CRUD API Refactor

### PR DESCRIPTION
**Pull request by kimberlyjk**
**Housekeeping**
- I moved around some files and re-factored a lot of my code to make the files more modular...
- I put the legacy database files into a new folder: `database/FEC` this files include `database.js`, which looks like a new file creation but infact is not. 
- i pulled the actual API functions out of `server/server.js` and move then into controller files that are located in `server/controllers`. There are two controller files, one for postgres and one for couchbase. There is also a new `helper.js` file that contains a bunch of helper functions that decipher the data from the DB and transforms it to client side, or vice versa. 
- I also pulled all the code that was inside `database/database.js` out and that file does the bare minimum, i.e. is opens the connection to the DB.

### **the Good Stuff**

`server/controllers/helper.js`
- as mentioned above this file has the functions that do the "heavy lifting" of converting the DB info into what the API Endpoint README specs. 

`database/cbDatabase.js`
- this file sets up the connections, etc. that will be needed when interacting witht couchbase _it should be noted that no code was refactored to actually interact/use couchbase in this PR_
- also should be noted that this file used to be called `couchbaseDB.js` and I decided to rename it so that my files would have a similar naming convention

`database/data_gen/dataGeneration.js`
- The way I was seeding the courseId =1 dataset was wrong. 

`database/database.js`
- not hard to see, but this file basically only opens up the connections to the databases and doesn't do much past that. 

`database/pgDatabase.js`
- this file has been renamed from `pSQLdatabase.js`
- It initalizes the Models and has the open and close connection functions that the `database.js` and `pgController.js` files will use to interact with the DB

`database/populateSQL.js`
- After struggling with getting the data generated into the postgres db to talk with the Models I had set up, I discovered that my lack of camelCase in "quotes" was failing me. so I refactored and reseeded my DB 

`server/controllers/cbController.js`
- a placeholder file for if I ever get around to writing the couchbase interactions for this service

`server/controllers/helper.js`
- this helper is a "two way" translator that will take info from the DB and convert it into client side format and vice versa. It takes different functions to do so, obviously, so it's a little hefty. 
- the ones that are labeled `transform` in the beginning are ones that are taking clientside inputs and manipulating the DB. Because of this, extra code has been written to ensure the formatting is as expected/as the DB requires. 

 `server/controllers/pgController.js`
- this is where the API endpoint functions live
- partially taken from the old `server//serverjs` file and partially refactored they have the API Endpoints specified in the [README](https://github.com/SDC-WebSockets/Sidebar#api)

`server/server.js`
- refactored similar to  `database/database.js` where most of the code has been removed and moved into a controller file for easier switching between pSQL and cb (if the latter were to ever occur)

and finally `test/sdc-api.spec.js`
- I moved the `DELETE` function before the `POST` because now we actually have a 101 courseId. 
- and I updated the previewVIdeo URLs to match what is expected.
- also updated the status codes to match what I had changed them to in the API endpoint

### CODE REVIEW

File: database/data_gen/dataGeneration.js [hunk 1]
```diff
@@ -79,7 +79,10 @@ const generatePriceData = async (numberOfCourses) => {
...
   const priceKeys = 'courseId,basePrice,discountPercentage,saleNumOfDays,saleOngoing';
   priceStream.write(`${priceKeys}\n`);
-  priceStream.write(`${course1.price}\n`);
+  const {
+    courseId, basePrice, discountPercentage, saleNumOfDays, saleOngoing,
+  } = course1.price;
+  priceStream.write(`${courseId},${basePrice},${discountPercentage},${saleNumOfDays},${saleOngoing}\n`);
 
   for (let i = 2; i <= numberOfCourses; i += 1) {
     const newPrice = createPricing(i);
```
Comment (mackenziekynoch) [inline database/data_gen/dataGeneration.js:91]: do you call writeStream.end() after your loop finishes? I believe the stream stays open until you explicitly close it or end the context where it's running

File: database/data_gen/dataGeneration.js [hunk 2]
```diff
@@ -109,7 +113,10 @@ const generateSidebarData = async (numberOfCourses) => {
   const sidebarStream = fs.createWriteStream(sidebarPath);
   const sidebarKeys = 'courseId,fullLifetimeAccess,assignments,certificateOfCompletion,downloadableResources';
   sidebarStream.write(`${sidebarKeys}\n`);
-  sidebarStream.write(`${course1.sidebar}\n`);
+  const {
+    courseId, fullLifetimeAccess, assignments, certificateOfCompletion, downloadableResources,
+  } = course1.sidebar;
+  sidebarStream.write(`${courseId},${fullLifetimeAccess},${assignments},${certificateOfCompletion},${downloadableResources}\n`);
 
   for (let i = 2; i <= numberOfCourses; i += 1) {
```
Comment (mackenziekynoch) [inline database/data_gen/dataGeneration.js:124]: what is the reason for looping 3 times instead of performing 3 actions (writing to each stream) in 1 loop?  Did you run into memory constraints with the latter approach?

File: database/database.js [hunk 3]
```diff
@@ -1,236 +1,11 @@
 /* eslint-disable no-restricted-syntax */
 /* eslint-disable no-console */
-const mongoose = require('mongoose');
 const dotenv = require('dotenv');
+const pg = require('./pgDatabase.js');
+// const cb = require('./cbDatabase.js');
```
Comment (mackenziekynoch) [inline database/database.js:5]: looks like you've got some leftover files from the inherited codebase / your earlier attempts.  Consider consolidating to only relevant files

File: database/pgDatabase.js [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
-const dotenv = require('dotenv');
 
-dotenv.config();
-const sequelize = new Sequelize('postgres://127.0.0.1:5432/SDC');
+const sequelize = new Sequelize('postgres://127.0.0.1:5432/SDC', {
+  benchmark: true,
+  logging: (sqlQuery, timing) => {
+    console.log(sqlQuery);
+    console.log('Query Time: ', timing, ' ms');
+  },
```
Comment (mackenziekynoch) [inline database/pgDatabase.js:11]: nice way to get the query time!

File: database/pgDatabase.js [hunk 5]
```diff
@@ -13,14 +18,19 @@ Price.init({
     unique: true,
     primaryKey: true,
   },
-  basePrice: DataTypes.DECIMAL,
+  basePrice: DataTypes.INTEGER,
```
Comment (mackenziekynoch) [inline database/pgDatabase.js:21]: LOOOVE that you moved your price to a reliable integer representation.  This is an excellent practice and also clearly demonstrates your experience and thoughtfulness with databases.  Definitely find a way to incorporate this into a Challenge-Action-Result narrative - it's very impressive!

File: database/pgDatabase.js [hunk 6]
```diff
@@ -13,14 +18,19 @@ Price.init({
...
   discountPercentage: DataTypes.INTEGER,
-  saleEndDate: DataTypes.DATE,
+  saleNumOfDays: DataTypes.INTEGER,
```
Comment (mackenziekynoch) [inline database/pgDatabase.js]: do you plan to update this daily as the end date grows closer? If you would prefer to store an integer, consider storing the millisecond representation of the date instead of the number of days from the time you create/update the record to the end date, since that requires a lot of upkeep

File: database/pgDatabase.js [hunk 7]
```diff
@@ -40,8 +55,10 @@ Sidebar.init({
     unique: true,
     primaryKey: true,
   },
-  fullLifetimeAccess: DataTypes.STRING,
-  accessTypes: DataTypes.STRING,
+  fullLifetimeAccess: {
+    type: DataTypes.BOOLEAN,
+    defaultValue: false,
+  },
```
Comment (mackenziekynoch) [inline database/pgDatabase.js:61]: great way to reduce the complexity and size of your db schemas with a simple byte-sized representation.  Another great callout as you show off your skills to future employers

File: server/controllers/helper.js [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    if (dbKeys.includes(currKey)) {
+      dbSidebar[currKey] = newSidebar[currKey];
+      if (currKey === 'fullLifetimeAccess') {
+        dbSidebar[currKey] = newSidebar[currKey] === 'Full lifetime access';
+      }
+    }
+  }
+  return dbSidebar;
+};
```
Comment (mackenziekynoch) [inline server/controllers/helper.js]: looks great overall.  You might consider making your function names a bit more descriptive so others can easily understand the work the function is doing without having to jump into this file to see

File: server/controllers/pgController.js [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+  const end = new Date();
+  const timeElapsed = end - start;
+  console.log('Time Elapsed: ', timeElapsed, 'ms');
+  if (errors.length > 0) {
+    res.status(400).send(errors);
+  } else {
+    res.send({ timeElapsed });
+  }
+  res.end();
+};
```
Comment (mackenziekynoch) [inline server/controllers/pgController.js:252]: nice job adding catches with helpful codes and errors

File: test/sdc-api.spec.js [hunk 10]
```diff
@@ -9,6 +9,8 @@ const app = require('../server/server');
 
 const request = supertest(app);
 
+beforeAll(() => jest.setTimeout(15 * 1000));
```
Comment (mackenziekynoch) [inline test/sdc-api.spec.js:12]: what's the reason for this?

File: test/sdc-api.spec.js [hunk 11]
```diff
@@ -40,8 +52,8 @@ describe('API calls', () => {
         downloadableResources: 20,
       },
       previewVideo: {
-        previewVideoImgUrl: 'http://localhost:3004/assets/previewVideoImg6.jpg',
-        previewVideoUrl: 'http://localhost:3004/assets/previewVideo6.mp4',
+        previewVideoImgUrl: 'videoImg/1.jpg#116378',
```
Comment (mackenziekynoch) [inline test/sdc-api.spec.js:55]: what is the hash at the end for?  Is that a normal convention?

**Review by mackenziekynoch — COMMENTED** [review]
Great job overall.  I'm so impressed with the decisions you made to make your db schema more memory efficient and more reliable - they clearly demonstrate mastery of schema concepts, so well done!  I added questions and thoughts all around, but the biggest callout which I'd recommend addressing is the 'saleNumOfDays' property in your data - if you're storing an integer as the difference between the day you create / update the record until the sale ends, you'll have to run a housekeeping update daily to ensure the sale ends on time.  If you would prefer to store that variable as an integer, consider storing the millisecond representation of the end date instead

File: database/data_gen/dataGeneration.js (hunk 1 above)
Comment (kimberlyjk) [reply database/data_gen/dataGeneration.js:91]: yep, I have a line on 91 for priceStream.end(); and sim for other writeStreams

File: database/data_gen/dataGeneration.js (hunk 2 above)
Comment (kimberlyjk) [reply database/data_gen/dataGeneration.js:124]: Yeah, i had memory issues when i ran them without awaiting for one to complete after the other.

File: server/controllers/helper.js (hunk 8 above)
Comment (kimberlyjk) [reply server/controllers/helper.js]: good call, I will change up the function names. I was struggling to figure out a good name without being too wordy and still not making sense.

File: test/sdc-api.spec.js (hunk 10 above)
Comment (kimberlyjk) [reply test/sdc-api.spec.js:12]: My tests were running before supertest had time to load up the server,etc. so the first GET test was always failing. I needed to add a setTimeout to avoid that.

File: test/sdc-api.spec.js (hunk 11 above)
Comment (kimberlyjk) [reply test/sdc-api.spec.js:55]: It is mostly to make the DB think that there are unique links so that it doesn't do any automatic caching. I think i just added it here so everything would look the same.

File: database/pgDatabase.js (hunk 6 above)
Comment (kimberlyjk) [reply database/pgDatabase.js]: this is a good point. When Jason was doing his DB he just added three days from day of data generation so i didnt think about that. I think I'll probably take your suggestion and save the end date as an integer in milliseconds.

File: database/database.js (hunk 3 above)
Comment (kimberlyjk) [reply database/database.js:5]: yep. will do. I am using the pg but not the dotenv. I dont know if I quite understand it so I left it in for now, but I don't think it's necessary in this file. I'll do that clean up.
